### ISSUE TITLE
Breadcrumb API is not thread-safe

### ISSUE THREAD

**Issue by dtorres** [bugsnag/bugsnag-cocoa#616]
### Description

While running our app under the thread sanitiser, it frequently hits a race condition on the Breadcrumb API as seen here:

<img width="444" alt="Screenshot 2020-05-22 at 11 09 20" src="https://user-images.githubusercontent.com/71180/82651636-b2bd7700-9c1c-11ea-8312-43957397edfb.png">
<img width="440" alt="Screenshot 2020-05-22 at 11 09 41" src="https://user-images.githubusercontent.com/71180/82651665-bf41cf80-9c1c-11ea-9b98-e3987d26f8fc.png">

### Environment

Library versions:

* Bugsnag version (from your Podfile, Podfile.lock, or elsewhere):
* CocoaPods version (if any) (`pod -v`):
* Carthage version (if any):
* iOS/tvOS/macOS version(s)
* debug mode or production?:
* How are you initializing Bugsnag? Can you share a snippet of your
  `AppDelegate` where Bugsnag is configured?

### Example code snippet

```swift
import Bugsnag

// (Insert code sample to reproduce the problem)
```

<details><summary>Error messages:</summary>

```

```
</details>

**Comment by mattdyoung** [comment]
Thanks for the report @dtorres. I can reproduce this so we'll take a look and try to determine the cause.

Have you seen any evidence of this causing crashes or other issues or has it just been flagged by thread sanitizer so far?

Until we can identify a fix, you may be able to work around this by using your own FIFO queue (e.g. `NSOperationQueue` with a `maxConcurrentOperationCount` count of 1) for calling `leaveBreadcrumb()`.

**Comment by dtorres** [comment]
> Thanks for the report @dtorres. I can reproduce this so we'll take a look and try to determine the cause.
> 
> Have you seen any evidence of this causing crashes or other issues or has it just been flagged by thread sanitizer so far?

We haven't seen this particular stack trace as a crash yet but a memory corruption issue (such as this) can cause other pieces of code to crash if they are unlucky enough to step on that same memory address later.

This race condition is now more likely to happen in our app as we have backgrounded most calls to the breadcrumb API as it isn't suited for main thread use (See #563).
 
> Until we can identify a fix, you may be able to work around this by using your own FIFO queue (e.g. `NSOperationQueue` with a `maxConcurrentOperationCount` count of 1) for calling `leaveBreadcrumb()`.

We are wrapping the calls inside of a lock to work around this.

**Comment by mattdyoung** [comment]
Understood. We'll need to investigate further on our side to determine the best way to fix. I'll update this issue when we know more.

**Comment by rob-keepsafe** [comment]
I saw this as well today and it's basically because you're writing a new breadcrumb within `readWriteQueue` but then not protecting reads for breadcrumbs within the same `readWriteQueue`. That violates the read/write approach that you all took since reads need equal protection

<img width="1783" alt="Screen Shot 2020-06-05 at 10 56 34 AM" src="https://user-images.githubusercontent.com/30269720/83903450-dba03a80-a6f9-11ea-8ef7-c869bd5f77a6.png">

<img width="1792" alt="Screen Shot 2020-06-05 at 10 57 00 AM" src="https://user-images.githubusercontent.com/30269720/83903541-0ab6ac00-a6fa-11ea-8c1e-d9faed6917ef.png">

**Comment by johnkiely1** [comment]
@rob-keepsafe 

Thanks for this, we will review and update this issue as necessary.

**Comment by RuiAAPeres** [comment]
Any luck on this?

**Comment by johnkiely1** [comment]
This has now been fixed in [V6.1.1 ](https://github.com/bugsnag/bugsnag-cocoa/releases/tag/v6.1.1)
